server/server.go

```
@@ -0,0 +1,170 @@
+package server
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/gorilla/mux"
+	log "github.com/sirupsen/logrus"
+	"github.com/yura-under-review/port-api/models"
+)
+
+const (
+	// MemoryLimit = 200 << 10 // 200 KB
+	MemoryLimit = 5 << 10
+)
+
+type Server struct {
+	addr             string
+	rootPageTemplate string
+	rootPageRendered []byte
+	s                *http.Server
+	repo             PortsRepository
+	sinkBatchSize    int
+}
+
+type PortsRepository interface {
+	UpsertPorts(context.Context, []*models.PortInfo) error
+}
+
+func New(config Config, repo PortsRepository) *Server {
+	return &Server{
+		addr:             config.Address,
+		rootPageTemplate: config.RootTemplateFile,
+		repo:             repo,
+		sinkBatchSize:    config.SinkBatchSize,
+	}
+}
+
+func (srv *Server) Run(ctx context.Context, wg *sync.WaitGroup) error {
+
+	// TODO: implement root page templating to setup host:port
+
+	var err error
+	srv.rootPageRendered, err = ioutil.ReadFile(srv.rootPageTemplate)
+	if err != nil {
+		log.Errorf("failed to read root page file: %v", err)
+		return err
+	}
+
+	r := mux.NewRouter()
+
+	r.HandleFunc("/", srv.rootHandler)
+	r.HandleFunc("/upload", srv.uploadHandler)
+
+	srv.s = &http.Server{
+		Addr:    srv.addr,
+		Handler: r,
+	}
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+
+		if err := srv.s.ListenAndServe(); err != http.ErrServerClosed {
+			log.Errorf("server failed: %v", err)
```

This error is not propagated